### ISSUE TITLE
Matrix profile (MPDist, motif discovery) applied to conversation patterns

### ISSUE THREAD

**Issue by codyschank** [TDAmeritrade/stumpy#149]
I have a data set of voice transcriptions from tech support calls, between customers and tech support agents. I've converted these transcripts into time series by doing the following: every time a word is said by the customer (+1), conversely every time a word is said by the agent (-1). I've attached a few examples of what these time series look like (each image is the time series described above for a single phone call between agent and customer). 

My initial goal was to cluster the calls based on the similarity of time series, and I discovered the KShape algorithm in my search for an appropriate method. In presenting this idea to a colleague, they suggested that I look into the Matrix Profile, which appears to be a very promising technique for analyzing these data. And that led me to discovering the MPDist algorithm. 

I used the STUMPY package to implement a basic MPDist function, but it is slow (~10 hours) to run across even a small set of these times series (n=500), due to all the pairwise comparison between time series that are needed, and the results did not appear to form interesting clusters (using HDBSCAN). One clear advantage of using MPDist over KShape is that it can handle time series of different lengths, which is a requirement for my data. 

I'm also interested in discovering motifs that are common across a set of these time series. I could see how to apply the motif discovery across a single time series, but applying the same tools to a set of time series was not clear to me (In this case I used an example from the matrixprofile-ts package). 

![6779281_timeseries_simple](https://user-images.githubusercontent.com/1004543/77367977-465fed80-6d29-11ea-8f02-d8550086ba23.png)
![7419837_timeseries_simple](https://user-images.githubusercontent.com/1004543/77367979-46f88400-6d29-11ea-94ce-02d2ecdb8371.png)
![8960944_timeseries_simple](https://user-images.githubusercontent.com/1004543/77367980-46f88400-6d29-11ea-81c2-2f97f3a93945.png)

**Comment by seanlaw** [comment]
@codyschank Thank you for posting this interesting application. I am by no means an expert on time series analysis but I certainly don't mind providing any suggestions that may come to my head (but know that they may also be bad suggestions too). To better understand, I have some questions for you:

1. What is the average/median number of words per phone call?
2. How many phone calls in total do you need to analyze and compare (pairwise)?
3. What kind of hardware are you currently using? What hardware do you have access to?
4. When you get a "high" MPdist value between two time series, have you tried visualizing them to see if the MPdist values actually make intuitive sense?
5. Can you share your MPdist code here?

IIRC, MPdist is a metric and not a distance measure and I vaguely remember the paper pointing out something related to triangular inequality (but I could be misremembering). 

If you see our [performance plot](https://github.com/TDAmeritrade/stumpy#performance), then you can get a ballpark estimate of the resources that you'll need to perform your computation. Not to speak ill of other implementations but our (likely biased) [timing calculations](https://github.com/TDAmeritrade/stumpy/issues/91) compared with matrixprofile-ts so that you'll get around a 5x speed up on a pair of modest CPUs with STUMPY and even more improvement with GPUs (a potential for 30x speedup). Unfortunately, GPUs is your best bet and it's fairly easy to install and run.

**Comment by seanlaw** [comment]
> I'm also interested in discovering motifs that are common across a set of these time series. I could see how to apply the motif discovery across a single time series, but applying the same tools to a set of time series was not clear to me (In this case I used an example from the matrixprofile-ts package).

Off the top of my head, I wonder if you could consider concatenating all of the calls into one long call. You may want to think about different strategies to reduce the size of the data by (+1 or -1) for every 5-10 words spoken by an agent/customer rather than for each individual word. That might get you into a regime where you have an order of magnitude less data to analyze (at least as a first pass). 

Additionally, I wonder if you could benefit from this [annotation approach](https://www.cs.ucr.edu/~eamonn/guided-motif-KDD17-new-format-10-pages-v005.pdf) published by Eamonn's group. Unfortunately, this is not implemented in STUMPY as I haven't had time to go through it thoroughly.

**Comment by seanlaw** [comment]
> I used the STUMPY package to implement a basic MPDist function, but it is slow (~10 hours) to run across even a small set of these times series (n=500), due to all the pairwise comparison between time series that are needed, and the results did not appear to form interesting clusters (using HDBSCAN). One clear advantage of using MPDist over KShape is that it can handle time series of different lengths, which is a requirement for my data.

Is there any way to perform a sanity check to make sure that this is a viable approach? I know Leland McInnes (co-creator of HDBSCAN) fairly well and he is also familiar with STUMPY so we may be able to ask him to chime in once we've done some due diligence. It may be possible that the calls that you've analyzed are just fundamentally different and that we are potentially searching for a signal that isn't there. I'm not trying to draw conclusions here but just trying to keep an open/skeptical mind. 😄

**Comment by seanlaw** [comment]
Btw, there is an [existing issue](https://github.com/TDAmeritrade/stumpy/issues/46) for adding an example of MPdist in the Tutorials. Would you be interested in taking a stab at it? We try to base our Tutorials on the original papers (i.e., we try to reproduce a figure from the original paper). This would be greatly appreciated and I am here to assist!

**Comment by codyschank** [comment]
> Btw, there is an [existing issue](https://github.com/TDAmeritrade/stumpy/issues/46) for adding an example of MPdist in the Tutorials. Would you be interested in taking a stab at it? We try to base our Tutorials on the original papers (i.e., we try to reproduce a figure from the original paper). This would be greatly appreciated and I am here to assist!

Definitely, I can help with that. I noticed that existing issue and thought about offering to help. Looks like I should be able to get the data here https://sites.google.com/site/mpdistinfo/

**Comment by codyschank** [comment]
> > I used the STUMPY package to implement a basic MPDist function, but it is slow (~10 hours) to run across even a small set of these times series (n=500), due to all the pairwise comparison between time series that are needed, and the results did not appear to form interesting clusters (using HDBSCAN). One clear advantage of using MPDist over KShape is that it can handle time series of different lengths, which is a requirement for my data.
> 
> Is there any way to perform a sanity check to make sure that this is a viable approach? I know Leland McInnes (co-creator of HDBSCAN) fairly well and he is also familiar with STUMPY so we may be able to ask him to chime in once we've done some due diligence. It may be possible that the calls that you've analyzed are just fundamentally different and that we are potentially searching for a signal that isn't there. I'm not trying to draw conclusions here but just trying to keep an open/skeptical mind. 😄

Let me know if you have any ideas for how to perform a sanity check. I'm also trying to approach this with skepticism and not just expecting that it will work on these data. I would expect there to be some patterns that repeat over and over, due to scripts that the agents follow and common problems that customers call in with. However, it does seem possible that the patterns would be different enough that the methods would not work to find clusters of similar calls. Your suggestion to plot calls with high and low MPDist seems like a good way to start investigating this.

**Comment by codyschank** [comment]
> > I'm also interested in discovering motifs that are common across a set of these time series. I could see how to apply the motif discovery across a single time series, but applying the same tools to a set of time series was not clear to me (In this case I used an example from the matrixprofile-ts package).
> 
> Off the top of my head, I wonder if you could consider concatenating all of the calls into one long call. You may want to think about different strategies to reduce the size of the data by (+1 or -1) for every 5-10 words spoken by an agent/customer rather than for each individual word. That might get you into a regime where you have an order of magnitude less data to analyze (at least as a first pass). This might help you quickly identify all of the top regions to start looking for motifs. And then, afterwards, you can go back to the full resolution to perform a higher resolution matrix profile calculation but that directly compares a prospective motif (plus or minus some additional data points) with the entire concatenated sequence.
> 
> Additionally, I wonder if you could benefit from this [annotation approach](https://www.cs.ucr.edu/~eamonn/guided-motif-KDD17-new-format-10-pages-v005.pdf) published by Eamonn's group. Unfortunately, this is not implemented in STUMPY as I haven't had time to go through it thoroughly.

Concatenating time series together does seem like a good approach. My expectation is that there are at least 3 general motifs, line going up (customer doing most of the speaking), line going down (agent doing most of the speaking), and a back and forth. The ultimate goal for me, aside from recognizing these motifs is to note where they occur in each call. The challenge I think (similar to the clustering) is fitting the "model" to a set of original data, and then applying the model as new data comes in without retraining (don't know if this analogy makes sense).

**Comment by seanlaw** [comment]
> Let me know if you have any ideas for how to perform a sanity check. I'm also trying to approach this with skepticism and not just expecting that it will work on these data. I would expect there to be some patterns that repeat over and over, due to scripts that the agents follow and common problems that customers call in with. However, it does seem possible that the patterns would be different enough that the methods would not work to find clusters of similar calls. Your suggestion to plot calls with high and low MPDist seems like a good way to start investigating this.

I'm a big proponent of looking at the individual raw time series that should "look" similar based on the MPdist value. Unfortunately, there is no substitute for this. 

Also, I'm wonder how much of a factor the (sliding) window size might play. Essentially, you'll have to figure out what an appropriate window size is for capturing a motif and this will require some domain expertise. Perhaps, the window size is, say, the median number of words needed (in total from both speakers) for the customer to have spoken twice or three times. Your guess is probably better than mine.

**Comment by seanlaw** [comment]
> Concatenating time series together does seem like a good approach. My expectation is that there are at least 3 general motifs, line going up (customer doing most of the speaking), line going down (agent doing most of the speaking), and a back and forth. The ultimate goal for me, aside from recognizing these motifs is to note where they occur in each call. The challenge I think (similar to the clustering) is fitting the "model" to a set of original data, and then applying the model as new data comes in without retraining (don't know if this analogy makes sense).

Yes, this makes sense. I sort of alluded to this above regarding the sliding window size. If you window size is too small then you won't capture "interruptions" or "changes" in speaker. If I understand correctly, "back-and-forth" would look like many small valleys and troughs. Is that correct? 

Not to dissuade you from using STUMPY but have you already tried applying simple statistics approaches to this? You may be able to segment each of your time series based on monotonically increasing/decreasing sections (i.e., only one person is speaking) and then ignore changes in that monotonicity that are too small or abrupt. 

Hmmm, I haven't thought this through thoroughly but I wonder if a Bayesian approach may be useful. I vaguely recall this [example from Cam Davison-Pilon's book](https://nbviewer.jupyter.org/github/CamDavidsonPilon/Probabilistic-Programming-and-Bayesian-Methods-for-Hackers/blob/master/Chapter1_Introduction/Ch1_Introduction_PyMC2.ipynb#Example:-Inferring-behaviour-from-text-message-data) and I'm thinking if you could build models this way. Again, this is completely unverified and may be an unnecessary rabbit hole :)

**Comment by codyschank** [comment]
> @codyschank Thank you for posting this interesting application. I am by no means an expert on time series analysis but I certainly don't mind providing any suggestions that may come to my head (but know that they may also be bad suggestions too). To better understand, I have some questions for you:
> 
> 1. What is the average/median number of words per phone call?
> 2. How many phone calls in total do you need to analyze and compare (pairwise)?
> 3. What kind of hardware are you currently using? What hardware do you have access to?
> 4. When you get a "high" MPdist value between two time series, have you tried visualizing them to see if the MPdist values actually make intuitive sense?
> 5. Can you share your MPdist code here?
> 
> IIRC, MPdist is a metric and not a distance measure and I vaguely remember the paper pointing out something related to triangular inequality (but I could be misremembering).
> 
> If you see our [performance plot](https://github.com/TDAmeritrade/stumpy#performance), then you can get a ballpark estimate of the resources that you'll need to perform your computation. Not to speak ill of other implementations but our (likely biased) [timing calculations](https://github.com/TDAmeritrade/stumpy/issues/91) compared with matrixprofile-ts so that you'll get around a 5x speed up on a pair of modest CPUs with STUMPY and even more improvement with GPUs (a potential for 30x speedup). Unfortunately, GPUs is your best bet and it's fairly easy to install and run.

1. About 2000 words on average
2. Similar to the answer about the motifs, I was thinking I would fit and identify clusters on a set of original data, and then predict which clusters new calls belong to as data streams in. I would guess somewhere around 5000 calls would be sufficient. But this is pretty much an arbitrary number. 500 seems too low (this is what I have been testing so far), 50000 is probably overkill. 
3. I'm currently using a desktop with 32 GB of RAM and Xeon 3.7 GHz processors.
4. I will do that as soon as I have the results again.
5. My function is pasted below (m is the window size, which I used as 120 seconds...I have been wondering if I need try different sizes). Reading the pseuocode [here](https://docs.google.com/viewer?a=v&pid=sites&srcid=ZGVmYXVsdGRvbWFpbnxtcGRpc3RpbmZvfGd4OjJmZGFkZWEwMTcxODYyMDM) it seems like I might have misunderstood the paper. 
```python
def MPDist(tsa,tsb,m):
    mp_ab = stumpy.stomp(T_A=tsa, T_B=tsb, m=m, ignore_trivial=False)
    mp_ba = stumpy.stomp(T_B=tsb, T_A=tsa, m=m, ignore_trivial=False)
    mp_abba = np.concatenate((mp_ab[:, 0],mp_ba[:, 0]))
    n = tsa.shape[0] + tsb.shape[0]
    k = int(round(2 * n * 0.05,0))
    MPdist  = np.sort(mp_abba)[k]
    return(MPdist)
```

You are almost right about the measure vs metric. A screen shot of the relevant passage from the paper is attached. Seems like maybe this might make MPDist a bad distance measure to use for clustering?

<img width="279" alt="measure vs metric" src="https://user-images.githubusercontent.com/1004543/77461928-53d0b280-6dd1-11ea-95f9-767fcc3642ea.PNG">

**Comment by codyschank** [comment]
> Not to dissuade you from using STUMPY but have you already tried applying simple statistics approaches to this? You may be able to segment each of your time series based on monotonically increasing/decreasing sections (i.e., only one person is speaking) and then ignore changes in that monotonicity that are too small or abrupt.

I have been wondering about this, and I'll probably start investigating this alternative soon. Perhaps calculating the slope of the time series within a moving window.

**Comment by seanlaw** [comment]
> My function is pasted below (m is the window size, which I used as 120 seconds...I have been wondering if I need try different sizes). Reading the pseuocode here it seems like I might have misunderstood the paper.

120 seconds or 120 words? You mentioned using transcripts but your plots above show time on the x-axis which then implies that your data takes into account pauses in talking?

The code looks right to me at first glance. What leads you to think that you may have misunderstood the paper?

> Seems like maybe this might make MPDist a bad distance measure to use for clustering?

Yes, this might be a problem but your application of HDBSCAN seems reasonable

**Comment by seanlaw** [comment]
Regarding a sanity check, you may even take a time series, `tsa`, and try adding a tiny bit of noise to it (call it `tsa_noise`). You can try:

1. Shifting the entire `tsa` up or down
2. Add noise (+1 or -1) randomly to 10%, 20%, 30% of `tsa`
3. Append additional randomly generated data to the end of `tsa` (i.e., extend the length of `tsa` by 10%, 20%, 30%)

And then see if any or all of these `tsa_noise` end up having an MPdist that is close to `tsa`

**Comment by codyschank** [comment]
> 120 seconds or 120 words? You mentioned using transcripts but your plots above show time on the x-axis which then implies that your data takes into account pauses in talking?

120 seconds. The x-axis is time in seconds, and takes into account pauses.

> The code looks right to me at first glance (and is comparable to [this](https://github.com/TDAmeritrade/stumpy/issues/46#issue-462691551) save for the sorting). What leads you to think that you may have misunderstood the paper?

The pseudo-code in the linked google site for MPDist lists MatrixProfile(TA, TB) as the only initial step before sorting and grabbing the value at k. This is what I did originally, but I ended up doing both MatrixProfile(TA, TB) and MatrixProfile(TB, TA) and concatenating the results, then sorting and grabbing the value at k. I did this because the Matrix Profile is not symmetric, and it appeared that it was right to do it the second way based on what as written out in the paper. I just need to reread that section more carefully. 

<img width="228" alt="symmetric" src="https://user-images.githubusercontent.com/1004543/77476372-791ceb00-6de8-11ea-8586-fb02840699f5.PNG">

**Comment by codyschank** [comment]
Found a bug in the MPDist function (but still not 100% sure I have it right according to the pseudocode). Previously I was assigning T_A to tsa and T_B to tsb for both of the first two lines. Rerunning things to see what kind of clusters I get. The function is corrected below.

```python
def MPDist(tsa,tsb,m):
    mp_ab = stumpy.stomp(T_A=tsa, T_B=tsb, m=m, ignore_trivial=False)
    mp_ba = stumpy.stomp(T_A=tsb, T_B=tsa, m=m, ignore_trivial=False)
    mp_abba = np.concatenate((mp_ab[:, 0],mp_ba[:, 0]))
    n = tsa.shape[0] + tsb.shape[0]
    k = int(round(2 * n * 0.05,0))
    MPdist  = np.sort(mp_abba)[k]
    return(MPdist)
```

**Comment by seanlaw** [comment]
This looks like the same code as you showed earlier above 🤔

**Comment by codyschank** [comment]
It's real small (and thus easy how I missed it).

First function:
```python
mp_ab = stumpy.stomp(T_A=tsa, T_B=tsb, m=m, ignore_trivial=False)
mp_ba = stumpy.stomp(T_B=tsb, T_A=tsa, m=m, ignore_trivial=False)
```

Second function:
```python
mp_ab = stumpy.stomp(T_A=tsa, T_B=tsb, m=m, ignore_trivial=False)
mp_ba = stumpy.stomp(T_A=tsb, T_B=tsa, m=m, ignore_trivial=False)
```

**Comment by seanlaw** [comment]
Now I see!

**Comment by codyschank** [comment]
Any thoughts about the most efficient/fastest way to apply this MPDist function across all pairs in an array? Basically I have a 1d array where each item is a time series.

**Comment by seanlaw** [comment]
Oh, no! I only noticed just now but you are using `stumpy.stomp()`! That is only there for reference purposes. Instead you should really be using `stumpy.stump()` (note STUMP and not STOMP). `stumpy.stump` is Numba JIT-compiled which means that it will leverage ALL of the cores on your machine. `stumpy.stomp` isn't even found in our [API documentation](https://stumpy.readthedocs.io/en/latest/api.html) and I should probably add a warning to users to switch. 

After that, if it is still not fast enough then you are running up against hardware constraints and, unfortunately, the only answer to your question is to use GPUs. If you already have GPUs installed on the same machine then simply changing your `stumpy.stump` functions to `stumpy.gpu_stump` should do the trick. As a back of the envelope comparison, something that takes around 10 hours on a 2 core machine would take around 20 minutes on a single 1080Ti GPU. If you don't have access to a GPU (and since your data isn't sensitive) then I recommend trying out Google Colab since they offer free GPUs. I just tested it out `gpu_stump` on there and it was pretty painless. Once you've spun up a notebook and imported your data there, be sure to turn on the GPUs via:

Edit > Notebook settings or Runtime>Change runtime type and select GPU as Hardware accelerator

Then:

```
!pip install stumpy

import numpy as np
import stumpy

x = np.random.rand(100000)
stumpy.gpu_stump(x, 50)
```

Let me know if that helps!

**Comment by seanlaw** [comment]
> 1. About 2000 words on average
> 2. Similar to the answer about the motifs, I was thinking I would fit and identify clusters on a set of original data, and then predict which clusters new calls belong to as data streams in. I would guess somewhere around 5000 calls would be sufficient. But this is pretty much an arbitrary number. 500 seems too low (this is what I have been testing so far), 50000 is probably overkill.

Based on this information, this implies that a concatenated sequence would be around `2000 x 500 = 1,000,000` in length. According to our performance metrics using `stumpy.stump` on this entire sequence would take around 2 hours on 2 modest commodity cores. This implies that `stumpy.stump` is around 5x faster than `stumpy.stomp`. With more cores, you should get roughly linear scaling. Also note that you can leverage Dask for distributed computing as well but the setup may be a bit cumbersome and I would advocate for GPUs first can have access.

@codyschank Would you mind reporting back what type of speed up you get from:

1. Switching to `stumpy.stump`
2. Using GPUs on Colab

I am really curious now!

**Comment by codyschank** [comment]
I'm wondering if I should implement numba on my for loop that applies MPDist across all pairs of an array of time series. Been trying it, but numba is new to me and I can't get it to work. Any obvious reason the code below would not work? I also attached a screen shot of time to run MPDist using stomp vs using stump, and stomp is faster. Could this be because my time series are relatively short? This is also why I'm wondering if I should apply numba to my for loop.

```python
@njit(parallel=True, fastmath=True)
def MPDist_loop(ts, m):
    n = len(ts)
    results = np.empty((n,n))
    for i in range(n):
        for j in range(n):
            if i != j:
                tsa = ts[i]
                tsb = ts[j]
                mp_ab = stumpy.stump(T_A=tsa, T_B=tsb, m=m, ignore_trivial=False)
                mp_ba = stumpy.stump(T_A=tsb, T_B=tsa, m=m, ignore_trivial=False)
                mp_abba = np.concatenate((mp_ab[:, 0],mp_ba[:, 0]))
                n2 = tsa.shape[0] + tsb.shape[0]
                k = int(round(2 * n2 * 0.05,0))
                MPdist  = np.sort(mp_abba)[k]
                results[i,j] = MPdist
    return(results)

<img width="250" alt="stomp vs stump" src="https://user-images.githubusercontent.com/1004543/77600001-90d19d80-6ed4-11ea-8157-2699c607867f.PNG">

```

**Comment by seanlaw** [comment]
> I'm wondering if I should implement numba on my for loop that applies MPDist across all pairs of an array of time series. Been trying it, but numba is new to me and I can't get it to work. Any obvious reason the code below would not work?

In your case, Numba doesn't actually help you since you're aren't actually performing any real computation (`stumpy.stump` is doing the computation and it's sent elsewhere). There's nothing really to parallelize. I think what you're really looking for is something like Python's `multiprocessing`. 

>  I also attached a screen shot of time to run MPDist using stomp vs using stump, and stomp is faster. Could this be because my time series are relatively short? This is also why I'm wondering if I should apply numba to my for loop.

Considering that your computations are taking milliseconds complete, this certainly suggests that your time series are too short. 😢

**Comment by seanlaw** [comment]
Usually with pairwise computations, it only makes sense to perform the upper/lower triangle like:

```
for i in range(n):
    for j in range(i+1, n):  # This is the change
        tsa = ts[i]
        tsb = ts[j]
        mp_ab = stumpy.stump(T_A=tsa, T_B=tsb, m=m, ignore_trivial=False)
        mp_ba = stumpy.stump(T_A=tsb, T_B=tsa, m=m, ignore_trivial=False)
        mp_abba = np.concatenate((mp_ab[:, 0],mp_ba[:, 0]))
        n2 = tsa.shape[0] + tsb.shape[0]
        k = int(round(2 * n2 * 0.05,0))
        MPdist  = np.sort(mp_abba)[k]
        results[i,j] = MPdist
        # Since you already have mp_ab and mp_ba, maybe you perform the reverse operations to get resulst[j, i]
```

**Comment by mexxexx** [comment]
Just as a quick note without having read everything you posted in detail:

The dev version of stumpy (cloning directly from github) supports `np.inf/np.nan` values in your timeseries, by skipping all subsequences that contain such a value. One trick I found in one of the MP papers (I think it was VI, about weakly labeled data), was to concatenate multiple time series into a large one, and having them separated by one of these illegal values. Then you can run `stump` to discover motifs that are present in some of the original time series you had, and because subsequences with illegal values are ignored you can be sure, that none of this motifs crosses any border.

I hope this makes sense, maybe it will help you somehow.

**Comment by seanlaw** [comment]
@mexxexx That's an excellent point! Thanks for chiming in. I will certainly keep that trick in mind.

I think that there is one subtlety in this case (but @codyschank please correct me if I'm wrong). Say, for 3 time series, he is interested in generating all of the pairwise matrix profiles. The problem with concatenating all 3 time series into one (even with an illegal value in between) is that it only returns the global matrix profile for each time series relative to the other two. However, I think what is needed is to find the matrix profile between all pairs time series. I would be very interested in a simple solution that could account for this

**Comment by mexxexx** [comment]
Well I cannot say anything about that, but if the goal is to only find patterns that are recurring in all of the calls, then the global matrix profile should be exactly what is needed.

**Comment by mexxexx** [comment]
> I also attached a screen shot of time to run MPDist using stomp vs using stump, and stomp is faster. Could this be because my time series are relatively short? This is also why I'm wondering if I should apply numba to my for loop.

Another reason is that stump has to be compiled first, while stomp is pure python code. So you have a lot of overhead there. When comparing speeds, you should always run numba compiled code before doing any comparisons to remove compile time from the equation. Maybe you could try it out and see if `stomp` is still faster, as that would be highly unexpected.

**Comment by codyschank** [comment]
> Well I cannot say anything about that, but if the goal is to only find patterns that are recurring in all of the calls, then the global matrix profile should be exactly what is needed.

Just wanted to point out that there are two use cases I'm trying to address, one is clustering using MPDist (the one I've focused on in most of these messages), but also finding motifs that are common across multiple time series and the trick suggested by @mexxexx seems like it will work perfectly. I will give it a try soon.

**Comment by codyschank** [comment]
Just wanted to give an update. 

1. I've set up my function to run in parallel using the multiprocessing library, so I can get results in a more reasonable amount of time (a dataset of 1000 time series takes about 1 - 2 hours). This has allowed me to do more experimenting.
2. I started using Spectral Clustering instead of DBSCAN, because I wanted to 'force' all of the time series into a cluster, rather than allowing some to be unclustered (as DBSCAN will do). 
3. MPDist was designed to use the value at k (2 * n * threshold) in the sorted matrix profile (now realizing I need to check if I sorted in the right order). I changed the function to use the mean value from the matrix profile. I have been wondering how this would change my results, but thought it might be more representative than selecting the value at k. 

My next step is to examine the clusters in detail to see if they are useful, but the visualizations seem promising.

**Comment by seanlaw** [comment]
That's great! Thanks for the update @codyschank

**Comment by seanlaw** [comment]
@codyschank Did you happen to read the [time series snippets paper](https://www.cs.ucr.edu/~eamonn/Time_Series_Snippets_10pages.pdf)? There might be some useful things in there for you as it relates to clustering and MPdist.

**Comment by tylerwmarrs** [comment]
@codyschank In practice, I have found local motifs (per time series) and clustered those with pure Euclidean distance to gain some insights. Generally speaking, I would extract any motif or anomaly found, store it in a database of labeled motif/discords and use clustering to identify if it has been seen before or not (based on a threshold). This is similar to the "most optimal" time series classification of 1NN classification matching. As your labeled data grows, the more accurate your predictions are.

Per @seanlaw on snippets, it only shows you the conservation of repeated patterns. In other words how much of a percentage does this pattern persist within my time series?

**Comment by tylerwmarrs** [comment]
@codyschank 
Not to pull you away from Matrix Profile clustering, but you may be interested in Dynamic Time Warping with lower bounding. Coincidentally, I have a repository showing implementation details, but it is kmeans clustering. My wife has used a similar implementation in practice successfully in sensor data space. https://github.com/tylerwmarrs/dtw-lbkeogh-clustering

The lower bounding algorithm is from Eamonn Keogh and the UCR group. @seanlaw you may find it interesting as well.

**Comment by codyschank** [comment]
> @codyschank
> Not to pull you away from Matrix Profile clustering, but you may be interested in Dynamic Time Warping with lower bounding. Coincidentally, I have a repository showing implementation details, but it is kmeans clustering. My wife has used a similar implementation in practice successfully in sensor data space. https://github.com/tylerwmarrs/dtw-lbkeogh-clustering
> 
> The lower bounding algorithm is from Eamonn Keogh and the UCR group. @seanlaw you may find it interesting as well.

@tylerwmarrs Does this approach work with time series of differing lengths? That is one challenge of my data set, and one reason I settled on using the matrix profile.

**Comment by tylerwmarrs** [comment]
It can. I am not sure if that implementation that I linked you to does. I know my wife had to adjust for different lengths.

**Comment by codyschank** [comment]
Just wanted to give another update, here's a quick look at the results from the clustering using Spectral Clustering based off on the MPDist

![image](https://user-images.githubusercontent.com/1004543/81113293-995ec000-8ee5-11ea-8e77-a7bce3f31b29.png)

I'm moving on to the motif discovery now. Got the dev version of stumpy installed, and calculated the matrix profile on a set of concatenated conversation time series with nans in between. However, I'm not sure the best way to automate motif discovery from the matrix profile. The best example I've seen is from the matrixprofile-ts libray. I'm trying replicate what they have done here, but let me know if you have any other ideas.

https://github.com/target/matrixprofile-ts/blob/master/matrixprofile/motifs.py

https://github.com/target/matrixprofile-ts/blob/master/docs/examples/Motif%20Discovery.ipynb

**Comment by seanlaw** [comment]
@codyschank Great stuff! Off the top of my head (and without looking at the matrixprofile-ts work), maybe you could (naively)

1. loop through the matrix profile
2. identify the index (from left to right) with the lowest matrix profile value to identify a motif
3. for the exclusion zone (i.e., +/- m/4) surrounding that index, set the matrix profile value to `np.inf`
4. go back to step 1. and repeat until you have as many motifs as you'd like

For example, if `m=4` and your matrix profile was:

`[6.9, 1.4, 6.2, 7.9, 11.4, 13.6, 14.1, 14.0, 1.4, 6.2]`

The first motif would be at `i=1` and then you'd update the the matrix profile to:

`[np.inf, np.inf, np.inf, np.inf, 11.4, 13.6, 14.1, 14.0, 1.4, 6.2]`

The next motif would be at `i=8` and then you'd update the matrix profile to:

`[np.inf, np.inf, np.inf, np.inf, 11.4, 13.6, np.inf, np.inf, np.inf, np.inf]`

And the final motif would be at `i=4` and the entire matrix profile would be updated to `np.inf`.

Does this work or get you in the ballpark? I could be wrong but I believe that the definition of a motif refers to the top/global matching pair (and it is a pair) while all others are just "nearest neighbors". In our example above, the first two are actually the global motifs and then everything else is a nearest neighbor. You may need to play around with how much to update the matrix profile to and you might want to make a copy first.

Btw, the `np.nan/np.inf` stuff should already be available in the released 1.3.1 version and can be installed with `python -m pip install stumpy` or `conda install -c conda-forge stumpy`. You no longer need to install from source!

**Comment by codyschank** [comment]
@seanlaw that makes sense for the most part. The only thing I am wondering is this: a particular motif should occur more than once, right? For your example, let's imagine the time series is longer and you just found the first occurrence of motif A, B, and C. How do I find other occurrences of those motifs in the time series?

**Comment by seanlaw** [comment]
@codyschank I think I see what you are asking. So, let's say using my method above, you find motif A first iteration through the loop and you get the index for it. Let's say the index is `i=17` Then, you go back to the output of `mp = stumpy.stump(T, m)` (which contains the matrix profile in the first column and the matrix profile index in the second column) and you look for `mp[17, 1]` and that should give you the index of the nearest neighbor to motif A (or "another occurrence of motif A". Of course, you'll want to check that the matrix profile distance to motif A is below some threshold.

Now, if you want to find all occurrences of a pattern (below some distance threshold) after you've identified the top 3 motifs, then you'll have to execute the `core.mass` algorithm. `core.mass` is "the fastest known algorithm" for taking a single subsequence and computing the distance to the entire time series and returning the distance profile.

```
import stumpy
import numpy as np

T = np.random.rand(1000) 
motif_a_idx = 17   # Assuming that you computed the matrix profile already with T and found a motif
stumpy.core.mass(T[motif_a_idx : motif_a_idx + m], T)
```

I apologize in advance if the last line isn't imported correctly as I am in the middle of updating `core.mass` code. Let me know if that makes sense

**Comment by codyschank** [comment]
Ok, I think I've got it. For the threshold, I'll follow the example from matrixprofile-ts. Looks like they use something called a radius (default=2), and threshold = distance * radius. Going back to your example a couple messages ago, the first motif has a value of 1.4 in the matrix profile, so the threshold would be 1.4 * 2 = 2.8 to find all other sequences that belong to that motif (as applied to the distance profile from stumpy.core.mass(T[motif_a_idx : motif_a_idx + m], T)). In matrixprofile-ts they also apply an exclusion zone to remove "trivial solutions". I will follow that as well. 

I'm trying to apply stumpy.core.mass, and I'm getting back mostly nans. Even though it seems like I'm calling correctly looking at https://github.com/TDAmeritrade/stumpy/blob/master/stumpy/core.py.

**Comment by seanlaw** [comment]
> Ok, I think I've got it. For the threshold, I'll follow the example from matrixprofile-ts. Looks like they use something called a radius (default=2), and threshold = distance * radius. Going back to your example a couple messages ago, the first motif has a value of 1.4 in the matrix profile, so the threshold would be 1.4 * 2 = 2.8 to find all other sequences that belong to that motif (as applied to the distance profile from stumpy.core.mass(T[motif_a_idx : motif_a_idx + m], T)). 

@codyschank I haven't looked at matrixprofile-ts but I'll offer a different perspective for the threshold as I don't think that setting it to `distance + 2` is a good idea. For example, if the majority of your distances are between 100-200, then setting the threshold to `distance + 2` would not be helpful. Instead, I would recommend maybe computing the standard deviation of the matrix profile (excluding NaNs/infs) and then using `distance + stddev(matrix profile)` as your threshold (or maybe some fraction of the standard deviation like 0.5 * stddev). In my opinion, your threshold should be set based upon your matrix profile.

> In matrixprofile-ts they also apply an exclusion zone to remove "trivial solutions". I will follow that as well.

So, since `core.mass` is one of those deep internal functions of STUMPY, a lot of the exclusion zone and `np.nan/np.inf` stuff is actually handled at a higher abstraction level. So, you'll have to perform some pre-processing steps first. What you may want is the `stamp.mass` function which is a nice wrapper around `core.mass` but it also handles the exclusion zone surrounding the reference/`trivial_idx`. So, you might do something like:

```
T = np.random.rand(1000)
m = 50

# Some pre-processing steps that is usually handled by `stumpy.stump`
T[np.isinf(T)] = np.nan
M_T, Σ_T = core.compute_mean_std(T_A, m)
T[np.isnan(T)] = 0
zone = excl_zone = int(np.ceil(m / 4))

# You compute the matrix profile and find a motif at `i = 10`
Q = T[10: 10 + m]
P, I = stamp.mass(Q, T, M_T, Σ_T, trivial_idx=10, excl_zone=zone)
```

**Comment by codyschank** [comment]
> @codyschank I haven't looked at matrixprofile-ts but I'll offer a different perspective for the threshold as I don't think that setting it to `distance + 2` is a good idea since `2` is completely arbitrary and is not meaningful in the absence of the matrix profile. For example, if the majority of your distances are between 100-200, then setting the threshold to `distance + 2` would not be helpful. Instead, I would recommend maybe computing the standard deviation of the matrix profile (excluding NaNs/infs) and then using `distance + stddev(matrix profile)` as your threshold (or maybe some fraction of the standard deviation like 0.5 * stddev). In my opinion, your threshold should be set based upon your matrix profile.

I've noticed in the comment above that you use `+` instead of `*`  That is a typo according to the formula I originally sent. Taking that into account, does that change your opinion of it being a good approach? 

 > ```
> T = np.random.rand(1000)
> m = 50
> 
> # Some pre-processing steps that is usually handled by `stumpy.stump` but you need to do this manually
> T_copy = T.copy()  # Leave original time series unchanged
> T_copy[np.isinf(T_copy)] = np.nan
> M_T, Σ_T = core.compute_mean_std(T_copy, m)
> T_copy[np.isnan(T_copy)] = 0
> zone = excl_zone = int(np.ceil(m / 4))
> 
> # You compute the matrix profile using `stumpy.stump` and ,say, you find a motif at `i = 10`
> Q = T_copy[10: 10 + m]
> P, I = stamp.mass(Q, T_copy, M_T, Σ_T, trivial_idx=10, excl_zone=zone)
> ```
> 
> Let me know if this works

Looks like P is just a single number, not the distance profile. I think this follows the discussion from issue #166. So it doesn't seem like the code will work as is. 

What I've done is just replace nan with 0 for the time series before I call `stumpy.core.mass` and this seems to work. I guess the only issue is that it will find motifs across the breaks, but I could remove those after the fact.

Some examples of motif sets found (scaled to all start at y = 0)

![image](https://user-images.githubusercontent.com/1004543/81732821-a5eb9700-9456-11ea-8bd8-b0438c22549c.png)

![image](https://user-images.githubusercontent.com/1004543/81732737-85bbd800-9456-11ea-96bd-0ed7b2c23456.png)

![image](https://user-images.githubusercontent.com/1004543/81732774-9409f400-9456-11ea-853d-fe3575e46cf9.png)

Hard to see on this figure, but this is whole series of 5 concatenated conversations, and where the motifs fall

![image](https://user-images.githubusercontent.com/1004543/81732977-e6e3ab80-9456-11ea-84ed-d7b7dfa36610.png)

**Comment by seanlaw** [comment]
> I've noticed in the comment above that you use + instead of * That is a typo according to the formula I originally sent. Taking that into account, does that change your opinion of it being a good approach?

Sorry for the typo. I still think that `distance * 2` still seems arbitrary and I would go with something like `distance + factor * stddev(mp)`. Maybe I'm misunderstanding but let's say that the majority of your matrix profile values were around 20 but your motif(s) had mp value(s) around 10 then you'd end up saying everything is a part of the same motif. Instead, if you computed the stddev(mp) and used `10 + stddev(mp)` then you'd eliminate a lot of other subsequences which, I think , is what you want. Again, I could be misunderstanding.

> Looks like P is just a single number, not the distance profile. I think this follows the discussion from issue #166. So it doesn't seem like the code will work as is.

Yes, unfortunately, `stamp.mass` needs to get renamed and doesn't quite achieve what you want. Your choice to use `core.mass` was correct!

> What I've done is just replace nan with 0 for the time series before I call stumpy.core.mass and this seems to work. I guess the only issue is that it will find motifs across the breaks, but I could remove those after the fact.

Yes, this would work

Regarding the discovered motifs: THAT'S SO COOL!!!  😄 Did you end up going the `distance * 2` route or the `distance + stddev(mp)` route?  I wonder if things will "look" better aligned if they were matched up by their mean or something like that? 

What do you think? How do you feel about the outcome of this analysis?

**Comment by seanlaw** [comment]
That last viz is super interesting! It's almost like you could do some sort of one-hot-encoding (or multi-hot encoding) for whether or not a motif exists in a given conversation. Kinda like:

```
Convo  Motif1  Motif2  Motif3  Motif4  Class Label
1      1       0       0       0       0
2      0       1       1       0       1
3      0       1       1       1       1
4      1       0       0       0       0
5      0       1       0       1       1
```

And then you can train an ML model to classify each call

**Comment by codyschank** [comment]
> Regarding the discovered motifs: THAT'S SO COOL!!! 😄 Did you end up going the `distance * 2` route or the `distance + stddev(mp)` route? I think the latter approach will produce aligned time series that are less noisy/deviating since the threshold would be smaller.
> I wonder if things will "look" better aligned if they were matched up by their mean or something like that?
> 
> What do you think? How do you feel about the outcome of this analysis? (I think you've done a wonderful job!)

Thanks! Couldn't have gotten this far without your help. I am very happy with the results. I used the approach from matrixprofile-ts for the threshold, not your approach (yet). I tried the analysis on a larger set of conversations (n =100) and all it gave back was a single motif set of complete silence, where the distance between the sequences of that set was zero. I think your approach will help solve this problem. Going to try it next. If not, I can think of a hack that might work.

**Comment by codyschank** [comment]
> And then you can train an ML model to classify each call. Is that what you were considering?

This is definitely one approach I was thinking about, but there are even simpler approaches that don't involve any modeling. For example, just knowing the placement of a particular motif within a conversation will be useful.

**Comment by codyschank** [comment]
I broke up long blocks of silence with nan values and got some really interesting results. This is for a set of 100 conversations. I also increased the radius to 3, and am still using the approach from matrixprofile-ts. Experimented a bit with your suggestion, but couldn't get good results. Probably just a matter of coming up with the right formula and parameter inputs. I'd like to try to read some papers that discuss motifs, and why the radius approach was developed (how it was tested). 

The other hack I tried for the problem with long blocks of silence, was masking out any matrix profile distances that were very close to zero. But that didn't work as well as breaking up the silence blocks with nans. 

About to test on an even larger dataset (n=1000). Thinking about how I can speed up the matrix profile calculation. For 100 conversations, took a minute or two. For 1000, we'll see. Perhaps I could get access to some GPUs.

![image](https://user-images.githubusercontent.com/1004543/81844240-64b9bc80-9514-11ea-9451-66343bafb255.png)

![image](https://user-images.githubusercontent.com/1004543/81844275-7307d880-9514-11ea-86d3-7ab386a459e0.png)

![image](https://user-images.githubusercontent.com/1004543/81844312-7dc26d80-9514-11ea-8c07-e220e7893743.png)

![image](https://user-images.githubusercontent.com/1004543/81844345-8dda4d00-9514-11ea-8d7d-170557865d6d.png)

![image](https://user-images.githubusercontent.com/1004543/81844374-99c60f00-9514-11ea-8545-e2db02609042.png)

![image](https://user-images.githubusercontent.com/1004543/81844397-a21e4a00-9514-11ea-9b17-f9c7fd5460b9.png)

![image](https://user-images.githubusercontent.com/1004543/81844436-b06c6600-9514-11ea-8511-6a7b2cb1b31d.png)

![image](https://user-images.githubusercontent.com/1004543/81844468-c24e0900-9514-11ea-8143-af7845ae3037.png)

![image](https://user-images.githubusercontent.com/1004543/81844497-cb3eda80-9514-11ea-936d-d8e1075bd9dd.png)

![image](https://user-images.githubusercontent.com/1004543/81844527-d560d900-9514-11ea-9f6e-0dfa830c6a7e.png)

**Comment by seanlaw** [comment]
> I broke up long blocks of silence with nan values and got some really interesting results. This is for a set of 100 conversations. I also increased the radius to 3, and am still using the approach from matrixprofile-ts. Experimented a bit with your suggestion, but couldn't get good results. Probably just a matter of coming up with the right formula and parameter inputs. I'd like to try to read some papers that discuss motifs, and why the radius approach was developed (how it was tested).

> The other hack I tried for the problem with long blocks of silence, was masking out any matrix profile distances that were very close to zero. But that didn't work as well as breaking up the silence blocks with nans.

These are some pretty convincing results! Have you also played around with different window sizes? Are the results still similar? When you say "Experimented a bit with your suggestion, but couldn't get good results", what would be a "bad" result? IIUC, you compute the motifs first (this is independent of the threshold) and then you go searching for other subsequences that are "close" to a given motif within some threshold. With a smaller/tighter threshold of `distance + stddev(mp)`, you'll just get less matches. Those matches shouldn't be "bad", they should be a subset of `distance * 2`. You could increase the `factor` in `distance + factor * (stddev(mp))` until `factor * (stddev(mp)) == 2 * distance` and you should get the same identical results as the `distance * 2` results. Maybe I'm missing something? Sorry, I don't mean to dwell. If you are finding what you are looking for then that's good enough for me!  👍 

> About to test on an even larger dataset (n=1000). Thinking about how I can speed up the matrix profile calculation. For 100 conversations, took a minute or two. For 1000, we'll see. Perhaps I could get access to some GPUs.

 As for GPUs, have you tried using Google Colab? Here's an [example Colab notebook](https://tiny.cc/stumpy_colab) that should help you get started. Note that I just added `nan/inf` support to `gpu_stump` in the development version (on Github) last night so you'll have to replace the `!pip install stumpy` command with something like `!pip install git+https://github.com/TDAmeritrade/stumpy.git` to get the latest version and it should be really fast. You'll just have to drop your unmarked data into Google Drive.

**Comment by codyschank** [comment]
> These are some pretty convincing results! Have you also played around with different window sizes? Are the results still similar?

Have not played around with the window size, but that is definitely something I will try.

> When you say "Experimented a bit with your suggestion, but couldn't get good results", what would be a "bad" result? IIUC, you compute the motifs first (this is independent of the threshold) and then you go searching for other subsequences that are "close" to a given motif within some threshold. With a smaller/tighter threshold of `distance + stddev(mp)`, you'll just get less matches. Those matches shouldn't be "bad", they should be a subset of those found using `distance * 2`. You could increase the `factor` in `distance + factor * (stddev(mp))` until `factor * (stddev(mp)) == 2 * distance` and you should get the same identical results as the `distance * 2` results. Maybe I'm missing something?
>
> Sorry, I don't mean to dwell and I don't want it to make it sound like I'm a narcissist/egomaniac who only thinks highly of his own ideas. I haven't put a ton of thought into the thresholding stuff and results reign supreme. So, if you are finding what you are looking for then that's good enough for me! 👍😄
>

Bad results for me are when no interesting motifs are returned (just blocks of silence for example). I think I might know what was going on now. There must be a bug with the max_motifs setting in that function. I never get back as many motifs as the max, always a couple less. When I was experimenting with your idea, I wasn't getting back very many motifs (just 1 or 2), and I must have had the max motifs set to 3 or 5 or something like that. I just experimented with increasing it to 50 in another situation and got back 48 really interesting motifs (this was on a data set of 50 conversations). I like your idea better. So I'm going to spend some more time working on it. 

> 
> As for GPUs, have you tried using Google Colab? Here's an [example Colab notebook](https://tiny.cc/stumpy_colab) that should help you get started. Note that I just added `nan/inf` support to `gpu_stump` in the development version (on Github) last night so you'll have to replace the `!pip install stumpy` command with something like `!pip install git+https://github.com/TDAmeritrade/stumpy.git` to get the latest version and it should be really fast. You'll just have to drop your unmarked data into Google Drive.

I think I have access to some GPU computing through work, so I'll try that first. But if it doesn't pan out, I will definitely try Google Colab.

**Comment by codyschank** [comment]
Quick update. I figured out how to run the algorithm without breaking up long silence blocks, and still get good results. @seanlaw  taking your approach with the threshold seems to have helped with that. I also figured out a way to deal with motifs crossing boundaries between two conversations. Currently, I've got some motifs that look like this 

![image](https://user-images.githubusercontent.com/1004543/83218555-bf1d6480-a133-11ea-98b1-006e12919bed.png)

Pretty good, except I would like the first 4 to really be in the same motif set. Essentially they are all silence, except for 1 - 5 words spoken. I just took a close look to see why these are not ending up in the same set. Basically, the distance between motif 1 (top left corner, all silence), and the next three in the first row is the max distance across the whole profile (10.95, matrix profile is below). This is a bit confusing to me. Also, the distance between any of motifs in sets 2 - 4 motifs is NaN. 

![image](https://user-images.githubusercontent.com/1004543/83218952-b5e0c780-a134-11ea-8d11-24e2c08083b6.png)

This is pretty much where I am stuck now. I could just combine the top 4 into one set manually, but I wonder if I am missing some other potential motifs due the weirdness I am seeing with the distances between these types of sequences (almost all silence). And it is one of the motifs I am most interested in.

**Comment by seanlaw** [comment]
That’s great. Thanks for the update!

**Comment by codyschank** [comment]
My hunch about some of the problems I see with the motifs (i.e. the 'flat' motifs falling in different motif sets), is that it had something to do with normalization when calculating the distances between subsequences. I did a quick google search and this article came up:

https://link.springer.com/chapter/10.1007/978-3-030-40014-9_5

This leads to me the question, is there a way to calculate the matrix profile in stumpy without doing normalization? 

Here is another example:

![image](https://user-images.githubusercontent.com/1004543/83288067-65a34d00-a1a8-11ea-81a0-2c2a3ccad3e9.png)

When normalized, the subsequences in this motif set would be very similar, but clearly they are very different in their raw form.

**Comment by seanlaw** [comment]
We have an [existing issue](https://github.com/TDAmeritrade/stumpy/issues/127) that discusses this but it might be a while before it can be prioritized. Would you be interested in contributing?

**Comment by codyschank** [comment]
Definitely! I've never contributed code to a package before, but definitely willing to give it a shot. What's the best way to get started? Clone the repo and start a new branch? 

I haven't forgotten that I also offered to help with the MPDist Example :)

**Comment by seanlaw** [comment]
> Definitely! I've never contributed code to a package before, but definitely willing to give it a shot. What's the best way to get started? Clone the repo and start a new branch?

> I haven't forgotten that I also offered to help with the MPDist Example :)

Absolutely no pressure but we welcome the help. But, yes, clone+branch. I actually like the idea of throwing things into a Jupyter notebook first and possibly hashing out the ideas there and getting feedback. Then, we can think about how to incorporate it later (or maybe not). In the [existing issue](https://github.com/TDAmeritrade/stumpy/issues/127), it sounds like @mexxexx may have some ideas. What I would say is to make sure to not get too attached to anything. What we don't want to do is end up needing to provide user support for anything that we aren't sure of or that not many people will end up using.

**Comment by mexxexx** [comment]
I think the referenced issue addresses a completely different issue, at least the implementation will be very different.

However, an unnormalized matrix profile might be interesting, but I can't say how difficult it is to implement, because the distance calculation would need to change quite a bit.

**Comment by seanlaw** [comment]
@codyschank Your application very loosely reminds me of [this article from a while ago](https://www.theatlantic.com/technology/archive/2016/07/the-six-main-arcs-in-storytelling-identified-by-a-computer/490733/). You may find inspiration in this story

**Comment by codyschank** [comment]
> I think the referenced issue addresses a completely different issue, at least the implementation will be very different.
> 
> However, an unnormalized matrix profile might be interesting, but I can't say how difficult it is to implement, because the distance calculation would need to change quite a bit.

I just read through the paper for removing noise from the matrix profile, and while it seems like it might help, I do think an un-normalized distance would probably do a better job with my use case.

I did a little more searching and this paper came up:

https://arxiv.org/abs/1901.05708

With an implementation by @tylerwmarrs here:

https://tylermarrs.com/posts/aamp-matrix-profile/

**Comment by codyschank** [comment]
I just now noticed this closed issue https://github.com/TDAmeritrade/stumpy/issues/157

So maybe this would not work.

**Comment by seanlaw** [comment]
> I just now noticed this closed issue #157

> So maybe this would not work. Though, it's not speed I'm worried about, but the results from the motif discovery based on the matrix profile.

I was just about to point out that issue and you beat me to it :)

**Comment by codyschank** [comment]
Despite the findings of @tylerwmarrs related to the AAMP algorithm, I am still very much focused on calculating a matrix profile based on non-normalized euclidean distance. I've been looking through the guts of the stumpy code, and today took my first stab at making changes. Seems to me like I would need to make the changes to ```stumpy.core._calculate_squared_distance()```. I've been looking for formulas that calculate raw euclidean distance also using the dot product, and came across this paper:

https://ieeexplore.ieee.org/document/8392419

And specifically, this section:

![image](https://user-images.githubusercontent.com/1004543/84696477-31eb5580-af12-11ea-8c4a-a6b341a79f72.png)

My thinking was the I could just substitute this formula for this one

```D_squared = np.abs(2 * m * (1.0 - (QT - m * μ_Q * M_T) / denom))```

The problem, I guess, is that I don't have sub-sequence A and B in ```stumpy.core._calculate_squared_distance()```. Before I go too deep down the rabbit hole, I wonder if you might have any opinion on if this is an approach that makes sense for what I'm trying to do. And if so, any recommendations for how to implement it.

**Comment by seanlaw** [comment]
@codyschank Would you mind if we started this conversation in a separate Github issue? This way, it is focused specifically on non-normalized Euclidean distance and it won't distract from this issue. I understand they are related to the same underlying problem that you are trying to solve but I would like to limit the scope of conversations so that it is easier to cross-reference in the future. I really appreciate it!

**Comment by codyschank** [comment]
Totally, I was wondering about that. This thread has grown rather long! I'll copy the last comment to a new issue.

**Comment by codyschank** [comment]
While working through #207 I had the realization that what would work well for this data is **not** a matrix profile based on non-normalized Euclidean Distance. Ideally, when the distance between two subsequences is calculated, they would both start at 0. Does it make sense why I would want to do this? Are you aware of any applications of the matrix profile that attempt to do this? I'm not even sure what to google in searching for an application like this.

**Comment by tylerwmarrs** [comment]
https://www.researchgate.net/publication/340663191_Neighbor_Profile_Bagging_Nearest_Neighbors_for_Unsupervised_Time_Series_Mining

@codyschank this might give you some inspiration.

**Comment by seanlaw** [comment]
> While working through #207 I had the realization that what would work well for this data is not a matrix profile based on non-normalized Euclidean Distance. Ideally, when the distance between two subsequences is calculated, they would both start at 0. Does it make sense why I would want to do this? Are you aware of any applications of the matrix profile that attempt to do this? I'm not even sure what to google in searching for an application like this.

@codyschank Sorry, I'm not following what you are saying here. Care to elaborate or provide a simple example?

**Comment by codyschank** [comment]
@seanlaw This is one example of a motif set using the regular (z-normalized) matrix profile. Note I have adjusted these subsequences to start at 0 when plotting them to make it easier to compare how well they match:

![image](https://user-images.githubusercontent.com/1004543/85893571-b6fb2800-b7b8-11ea-85f8-6efe6d3ae50e.png)

I don't really like that the orange subsequence gets lumped with the red subsequence. The magnitude of the shapes is very different (approx -40 vs -3). It is the z-normalization of the matrix profile that puts them together. 

Here are the same subsequences without me plotting them all starting at zero. 

![image](https://user-images.githubusercontent.com/1004543/85893812-2ffa7f80-b7b9-11ea-8c5e-3c1219820884.png)

If I used the non-normalized Euclidean distance, orange and purple would have one of the larger distances, even though they are quite similar if you plot them starting from zero. 

The best result for me is to have green, brown, and red in one motif, and orange, purple, and blue in another. 

The example is not perfect, but I hope it gets the point across. It is one that could provide quickly.

**Comment by seanlaw** [comment]
@codyschank Ahhh, got it. Thank you. I will think about this.

**Comment by seanlaw** [comment]
@codyschank Just wanted to touch base and see how you were doing with this? I finally had some time to read through the AAMP paper and I might have some time to take a stab at implementing the AAMP approach. Of course, I don’t want to compete and waste your time and effort either. Please let me know what you think.

**Comment by seanlaw** [comment]
@codyschank Do you mind if we close this for now? And feel free to start a new issue (referencing this one) if there are follow-ups? Thanks!

**Comment by codyschank** [comment]
Sounds good to me. I'll close it.
